src/server/analytic_services/prometheus_reporting.js

```diff
@@ -1,485 +1,115 @@
-                    this._metrics.object_histo.set({ size: cur.label, avg: cur.avg }, cur.count, new Date());
-                }
-            }
+    const server = http.createServer((req, res) => {
+        // Serve all metrics on the root path.
+        if (req.url === '' || req.url === '/') {
+            res.writeHead(200, { 'Content-Type': 'text/plain' });
+            res.end(export_all_metrics());
+            return;
         }
-    }
 
-    set_providers_bandwidth(type, write_size, read_size) {
-        if (!this.enabled()) return;
-        this._metrics.providers_bandwidth_read_size.set({ type }, read_size);
-        this._metrics.providers_bandwidth_write_size.set({ type }, write_size);
-    }
-
-    set_providers_ops(type, write_num, read_num) {
-        if (!this.enabled()) return;
-        this._metrics.providers_ops_read_num.set({ type }, read_num);
-        this._metrics.providers_ops_write_num.set({ type }, write_num);
-    }
-
-    set_providers_physical_logical(providers_stats) {
-        if (!this.enabled()) return;
-        for (let [type, value] of Object.entries(providers_stats)) {
-            const { logical_size, physical_size } = value;
-            this._metrics.providers_physical_size.set({ type }, physical_size);
-            this._metrics.providers_logical_size.set({ type }, logical_size);
+        // Serve report's metrics on the report name path
+        const report_name = req.url.substr(1);
+        if (reports.has(report_name)) {
+            res.writeHead(200, { 'Content-Type': 'text/plain' });
+            res.end(export_report_metrics(report_name));
+            return;
         }
-    }
 
-    set_system_info(info) {
-        if (!this.enabled()) return;
-        this._metrics.system_info.reset();
-        this._metrics.system_info.set({ system_name: info.name, system_address: info.address }, Date.now());
-    }
+        res.writeHead(404, { 'Content-Type': 'text/plain' });
+        res.end();
+    });
 
-    set_system_links(links) {
-        if (!this.enabled()) return;
-        this._metrics.system_links.reset();
-        this._metrics.system_links.set({ resources: links.resources, buckets: links.buckets, dashboard: links.dashboard }, Date.now());
-    }
-
-    set_bucket_status(buckets_info) {
-        if (!this.enabled()) return;
-        this._metrics.bucket_healthy.reset();
-        this._metrics.bucket_quota_precent.reset();
-        this._metrics.bucket_capacity_precent.reset();
-        buckets_info.forEach(bucket_info => {
-            this._metrics.bucket_healthy.set({ bucket_name: bucket_info.bucket_name }, Number(bucket_info.is_healthy));
-            this._metrics.bucket_quota_precent.set({ bucket_name: bucket_info.bucket_name }, bucket_info.quota_precent);
-            this._metrics.bucket_capacity_precent.set({ bucket_name: bucket_info.bucket_name }, bucket_info.capacity_precent);
-        });
-    }
+    return new Promise((resolve, reject) => {
+        server.listen(port, err => {
```

Comment:
Maybe we should handle the failure and attempt to bring it up later on? After a delay or something? 
Comes back to my other comments regarding waiting and handling failures in different processes.

Reply:
I do not agree,
If the port is taken it will not suddenly be available. 
If it will we have an issue we need to deal with and not just hide it be "trying later"

Reply:
So deal with it instead of trying later.
I do not see you dealing with it.
The current state is that you do not deal with it and run without metrics.

Reply:
I understand the argument about time constraint and not wanting to handle errors properly now, that is valid and I agree with it.
However, I don't want a one try here, which we know would fail at some point. I don't think we know enough and not it's not just about port taken and then being avail. 

Please add a retry mechanism, we can do 10s between each retry and we can limit on, say 30 such retries (which would account for an entire aggregation cycle). We should also log in high severity that we failed between retries. That should not consume much time to add.

Reply:
@nimrod-becker I highly disagree with that:

A retry in any case that should not fail (like opening a port) will only hide a real problem.
Think about a race condition between agents (not noobaa agent, conceptual ones) that try to open the same port.
The best caser scenario, with respect to support and debugging, is that the issue can be reproduced in a consistent manner. Adding a retry loop can and will hide such collisions if the other side is also doing the same. 

All of the above will lead to unreproducible bugs, on even to 2 different runs on the same machines that yield different results! this is very bad for us, for QE, and for the customer.

In my opinion, a retry should be added only if we know that an action would fail (not should) and that the situation is temporary (like connectivity to the cloud). 
I cannot say this is the case when trying to start a server and you can even see that we took the same decision with our  HTTP/HTTPS servers (see endpoint code).

We should handle edge cases (error handling)  when we will encounter them (not upfront). Then if we have a situation we know how to handle,  a retry will be a good solution after the error handling.

Saying all of the above,
Let leave it as is

Reply:
@nb-ohad 
```
It is a different case altogether, in the agent case you are trying to establish a connection here you are trying to listen on a port. Trying to establish a connection might be affected by network hiccups that might resolve with a retry.
A retry IS THE FIX in that case but not in this one.
```
So you are saying there couldn't be any hiccups that can be resolved by a retry when you attempt to listen to a port?

```
Other failures are persistent and will fail in every retry.
Retry or not it will not make a difference.
```
The same way as establishing connections can be persistent and cannot be fixed by retrying.
And I do not agree with that. Listening on ports can be non-persistent issues.

```
Trying to listen on a port by more than one actor is not persistent and prune to races.
Total network issues, like problematic network adaptors, will fail our actual server not only the metrics server.
```
```
Trying to listen on a port by more than one actor is not persistent and prune to races.
Total network issues, like problematic network adaptors, will fail our actual server not only the metrics server.
```
You keep on locking down to multiple processes that listen on the same port.
Obviously it won't fail on the metrics server, because you wouldn't deploy it since you give up immediately.
There could be problems with acquiring ports on certain machines that can be fixed and retried and are not persistent, already had customers with such scenarios.
You keep on stating that port listening is a persistent problem that can only happen if someone is acquired that port and if not then it is not our problem.

```
Under the assumption that there are bugs in the product, ask a customer if he/she prefers a product with a hidden bug that a retry might be or not be a workaround but with longer support time and a "half" ability to diagnose or a product that will fail consistently but with a better diagnose and response time from support.
```
First of all, unknown bugs are unknown... So there is not much said there. Let's assume that there are bugs.
Let's gather customers and ask them then (I assume not feasible, but we can try).
I do not see why adding a retry will cause "longer support time and a "half" ability to diagnose or a product that will fail consistently".
Please explain how adding retry will cause: a longer cycle of support, make the product not fail consistently.
If the customer has a consistent problem then it will fail consistently with the retry and without.
If the problem is not consistent then it will not fail consistently with and without the retry.
A longer cycle of support because of adding a retry? Why? You have the same debug logs and information in both cases (with and without retry).

```
If you think metrics are so valuable you should advocate that a failure to start the metrics server should be a service failure and the service should be stoped. I don't believe that you believe that, but please correct me if I am wrong.
```
I believe that metrics are valuable enough for us to try and redeploy them several times instead of throwing our hands on the first failure. I'm not in any way saying that the failure to deploy a metrics server should block the working of NooBaa, all I'm saying that we shouldn't give up on the first try.

```
Two options:

1) The retry always works - you will never know of the underlying bug.
2) The retry might work but not always, reproducing the problem and getting consistent results can be very problematic, and because this might happen a long time from now, the chance that we will immediately understand that we have retries that mask the problem is almost non-existent, we will waste a long time in just trying to reproduce and diagnose why it is happening only sometimes.
```
1) The retry always works - you have the same information as you did without the retry.
2) If the problem is consistent you have consistent results with or without retry. You will understand that we have retries because you can read the code (if you want you can add the retry number in the error print). 
Reproduction is the same for both cases with or without the retry. You will need to diagnose if it happens only sometimes both with and without the retry (hence the persistent or non-persistent error).

```
The underlaying this here is that this case does not fall under the category of "retry as a solution" (like establishing a connection to a server) but into the category "retry as a workaround" and a workaround should be employed after you have a specific failure case and in that specific case, you want to dim a retry as a viable workaround. not doing so and applying a retry loop to all cases will hide bigger problems that should be addressed.
```
I'm not quite sure why you consider retrying agent initial connection as a "solution", yet retry on initial listening on ports as a "workaround".
We don't even deal with the specific error on agent initial connect and then retry, it is literally the same.
By your logic of "workaround" and "solution", there shouldn't be any error handling in code unless you know all of the possible errors fix them, and then retry.
Did I understand you correctly?

Reply:
@jeniawhite 
This conversation does not go anywhere and I did not find any new information or insight in your last post that would convince me otherwise.
 
The bottom line and my general belief are that we should always strive (even if it is not always possible) for a system that behaves consistently. 

If something should fail for an unknown reason it should always fail for that reason and I don't care what it is.
if we know of a potential reason and decide to fortify the product against it then we should do it, even if we decide that for that specific error the solution is a retry.

But what we should not do is play with dices, it will create a system that behaves unexpectedly in all cases. These kinds of systems are hard to diagnose and debug. And the turn-around time for solution is always bigger in such systems.

Reply:
@nb-ohad 
We've already had another point of view from @nimrod-becker 
We should bring it up with other team members in addition to @guymguym , as well then.

Reply:
@jeniawhite @nb-ohad You've summoned me. 
What can I do for you? You have 3 wishes (per PR), use them wisely :)

Reply:
3 Wishes are enough :) We need 1 I think.

Jenia and I seem to think we better handle a potential error on the listen in the form of a few more retries (not forever, we can decide how many and how much backoff).

Ohad raises the point that we better fail immediately since its both very unexpected to fail to listen and it can mask the original problem (Ohad I hope I did not misrepresent your opinion)

Reply:
@guymguym a better description of my point of view can be summarized in my last comment above

Reply:
OK. I've taken a close look.

I think the theological discussion, while important, is hard on this platform, so I would encourage you to raise such topics in the next community meeting - just add it to the agenda and take 10 minutes to discuss about it.

I think if we enabled this feature and it fails to start up correctly, we should exit.
See how this is done in web server (regardles that we should deprecate the use of P.ninvoke)
https://github.com/noobaa/noobaa-core/blob/22de4be4bdf1b0cfac7d89dda371f67e6e178264/src/server/web_server.js#L91-L94

Reply:
To sum up yesterday's discussion

1) we would add 5 retries with 20s between each retry, print each failure to the log on high sev
2) we would not fail the server since metrics unavailability (even for dashboards and optimization) would be better then data unavailability if core can't come up

Reply:
After the discussion (slack) with I did the following changes. 
1. will retry a fixed number of times based on config/env, logging each failure, with a  delay based on config/env.
2. will exist the process after all retries 

Default values are:
- Infinity for retry_count.
- delay is 5 min